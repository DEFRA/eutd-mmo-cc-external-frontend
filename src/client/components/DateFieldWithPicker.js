import React from 'react';
import { DateField, LabelText } from 'govuk-react';
import DatePicker from 'react-datepicker';
import moment from 'moment';
import { withTranslation } from 'react-i18next';

export default withTranslation() (class DateFieldWithPicker extends React.Component {
  state = {
    showDatePicker: false,
    selectedDate: parseDate(new Date()),
    dateObject: {
      day: '',
      month: '',
      year: '',
    },
  };

  componentDidUpdate(prevProps) {
    const { date, dateFormat } = this.props;
    if (prevProps.date !== date && date !== 'Invalid date') {
      this.setDateValue(date, dateFormat);
    }
  }

  setDateValue = (date, dateFormat) => {
    const formattedDate = moment(date, [dateFormat]);
    if (date) {
      this.setState({
        selectedDate: formattedDate,
        dateObject: {
          day: formattedDate.format('DD'),
          month: formattedDate.format('MM'),
          year: formattedDate.format('YYYY'),
        },
      });
    } else {
      this.setState({
        selectedDate: null,
        dateObject: {
          day: '',
          month: '',
          year: '',
        },
      });
    }
  };

  handleOnSelect = (date) => {
    const checkDate = moment(date);
    this.props.onDateChange({
      target: {
        name: this.props.name,
        value: moment(date, ['YYYY-M-D', 'YYYY-MM-DD'], true).format(
          this.props.dateFormat
        ),
      },
    });
    this.setState({
      showDatePicker: false,
      selectedDate: date,
      dateObject: {
        day: checkDate.format('DD'),
        month: checkDate.format('MM'),
        year: checkDate.format('YYYY'),
      },
    });
  };

  componentDidMount() {
    const { date, dateFormat } = this.props;
    if (date !== 'Invalid date') {
      this.setDateValue(date, dateFormat);
    }
    ['dayInputName', 'monthInputName', 'yearInputName'].forEach((name) => {
      const inputEl = Array.from(document.getElementsByName(name) || [])[0];

      if (inputEl) {
        inputEl.addEventListener('keydown', (event) => {
          (event.keyCode === 69 || event.keyCode === 190) &&
            event.preventDefault();
        });
      }
    });
  }

  onDateChange = (value) => {
    this.setState({
      dateObject: value,
    });
    const { dateFormat, name } = this.props;

    if (value && Object.values(value).every((_) => _ !== '')) {

      this.props.onDateChange({
        target: {
          name: name,
          value: getFormattedDate(value, dateFormat),
        },
      });
      this.setState({
        selectedDate: parseDate(
          value.day + '/' + value.month + '/' + value.year
        ),
      });
    } else {
      this.setState({ selectedDate: null });
      this.props.onDateChange({
        target: {
          name: name,
          value: 'Invalid date',
        },
      });
    }
  };

  render() {
    const { showDatePicker, selectedDate, dateObject } = this.state;
    const { errors, labelText, labelTextClass, id, t} = this.props;

    return (
      <>
        <div className="date-field-with-picker" id={id}>
          <div className="column-field-picker">
            <DateField
              className="date-field"
              hintText={t('psAddHealthCertificateDateFieldHint')}
              errorText={errors}
              inputNames={{
                day: 'dayInputName',
                month: 'monthInputName',
                year: 'yearInputName',
              }}
              input={{
                labels: {
                  day: t('commonDatePickerLabelDay'),
                  month: t('commonDatePickerLabelMonth'),
                  year: t('commonDatePickerLabelYear')
                },
                value: {
                  day: dateObject.day || '',
                  month: dateObject.month || '',
                  year: dateObject.year || '',
                },
                onChange: this.onDateChange,
              }}
            >
              <LabelText id="date-field-label-text" className={labelTextClass}>
                {labelText}
              </LabelText>
            </DateField>
            <div className={`date-picker ${errors ? 'error' : ''}`}>
              <button
                onClick={(e) => {
                  e.preventDefault();
                  this.setState({ showDatePicker: true });
                }}
              >
                <img alt="calendar" src="/static/assets/images/calendar.png" />
              </button>
            </div>
          </div>
          {showDatePicker && (
            <DatePicker
              inline
              withPortal
              autoComplete="off"
              selected={selectedDate || ''}
              dateFormat={'DD/MM/YYYY'}
              onClickOutside={() => this.setState({ showDatePicker: false })}
              onSelect={this.handleOnSelect}
            />
          )}
        </div>
      </>
    );
  }
});

export function getFormattedDate(value, dateFormat) {

  if (dateFormat.startsWith('YYYY')) {
    const dateString = value.year + '-' + value.month + '-' + value.day;
    return moment(dateString, ['YYYY-M-D', 'YYYY-MM-DD'], true).format(
      dateFormat
    );
  } else {
    const dateStringReverse = value.day + '/' + value.month + '/' + value.year;

    return moment(dateStringReverse, ['D/M/YYYY', 'DD/MM/YYYY'], true).format(
      dateFormat
    );
  }
}

export function parseDate(date) {
  const b = date && moment(date, 'DD/MM/YYYY');
  if (!b || !b.isValid()) return null;
  return b;
}
