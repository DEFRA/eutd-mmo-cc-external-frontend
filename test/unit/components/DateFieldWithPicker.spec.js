import { mount } from 'enzyme';
import * as React from 'react';
import moment from 'moment';
import DateFieldWithPicker, { parseDate, getFormattedDate } from '../../../src/client/components/DateFieldWithPicker';
import { act } from 'react-dom/test-utils';
import { render } from '@testing-library/react';


describe('DateFieldWithPicker', () => {

  const props = {
    errors: undefined,
    onDateChange: jest.fn(),
    labelText: '',
    labelTextClass: 'label-landings-form',
    dateFormat:'YYYY-MM-DD',
    id: 'component-id'
  };
  const wrapper = mount(<DateFieldWithPicker {...props} />);

  afterEach(() => {
    jest.resetAllMocks();
  });

  it('Should render a valid react component', () => {
    expect(wrapper).toBeDefined();
  });

  it('Should render a DateField', () => {
    expect(wrapper.find('DateField').exists()).toBeTruthy();
  });

  it('Should render a Label text field with the correct className', ()=> {
    expect(wrapper.find('#date-field-label-text').first().text()).toEqual(props.labelText);
    expect(wrapper.find('#date-field-label-text').first().props().className).toContain(props.labelTextClass);
  });

  it('Should render a DatePicker', () => {
    expect(wrapper.find('div.date-picker').exists()).toBeTruthy();
  });

  it('Should set showDatePicker to true when the calendar icon is clicked', () => {
    const calendar = wrapper.find('div.date-picker');
    expect(calendar).toBeDefined();
    calendar.find('img').simulate('click', { preventDefault: () => { } });

    expect(wrapper.find('DatePicker').exists()).toBeTruthy();
    expect(wrapper.find('CalendarContainer.react-datepicker').exists()).toBeTruthy();
    expect(wrapper.find('DateFieldWithPicker').state('showDatePicker')).toBe(true);
  });

  it('Should set selected date when a date is selected', () => {
    const calendar = wrapper.find('div.date-picker');
    expect(calendar).toBeDefined();
    calendar.find('img').simulate('click', { preventDefault: () => { } });
    const datePicker = wrapper.find('DatePicker');
    act(() => {
      datePicker.prop('onSelect')({ day: '01'});
    });
    expect(wrapper.find('DateFieldWithPicker').state('selectedDate')).toEqual({day: '01'});
  });
  
  it('Should have an id as props', () => {
    expect(wrapper.props().id).toEqual('component-id');
  });

  it('it should create snap shot for whole page', () => {
    const { container } = render(wrapper);
    expect(container).toMatchSnapshot();
  })
});

describe('DatePicker callback props', () => {
  const props = {
    errors: undefined,
    onDateChange: jest.fn(),
    dateFormat:'YYYY-MM-DD'
  };
  const wrapper = mount(<DateFieldWithPicker {...props} />);
  let datePicker;
  beforeEach(() => {
    wrapper.find('img').simulate('click', { preventDefault: () => {} });
    datePicker = wrapper.find('DatePicker');
  });

  it('Should set showDatePicker to false when clicking outside', () => {
    datePicker.instance().props.onClickOutside();
    expect(wrapper.find('DateFieldWithPicker').state('showDatePicker')).toBe(false);
  });

  it('will set the right state when the picker is selected', () => {

    const date = { day: '01', month: '05', year: '2021' };
    const checkDate = moment(date);

    datePicker
      .instance()
      .props.onSelect(date);
    expect(wrapper.find('DateFieldWithPicker').state()).toEqual({
      dateObject: { day: checkDate.format('DD'), month: checkDate.format('MM'), year: checkDate.format('YYYY') },
      selectedDate: {
        day: '01',
        month: '05',
        year: '2021',
      },
      showDatePicker: false,
    });
  });
});

describe('DateField', ()=> {
  let wrapper;
  const mockAddEventListener = jest.fn();
  const mockOnDateChange= jest.fn();
  const props = {
    onDateChange: mockOnDateChange,
    errors: undefined,
    dateFormat:'YYYY-MM-DD'
  };

  beforeEach(() => {
    jest.spyOn(document, 'getElementsByName')
    .mockReturnValue([{
      addEventListener: mockAddEventListener
    }]);

    window.scrollTo = jest.fn();
    wrapper = mount(<DateFieldWithPicker {...props} />);
  });

  afterEach(() => {
    jest.resetAllMocks();
  });

  it('should update onChange every time the user types in the date field', () => {
    const dayField = wrapper.find('input[name="dayInputName"]');
    const monthField = wrapper.find('input[name="monthInputName"]');
    const yearField = wrapper.find('input[name="yearInputName"]');

    dayField.simulate('change', { target: { value: '10' } });
    monthField.simulate('change', { target: { value: '07' } });
    yearField.simulate('change', { target: { value: '2021' } });

    expect(wrapper.find('DateFieldWithPicker').state()).toMatchObject({
      showDatePicker: false,
      dateObject: {
        day: '10',
        month: '07',
        year: '2021',
      },
    });
  });

  it('should add an event listener to prevent alpha characters', () => {
    expect(mockAddEventListener).toHaveBeenCalledTimes(3);
  });

  it('will set selectedDate to be null if the date in the field is not finished', ()=> {
    const dayField = wrapper.find('input[name="dayInputName"]');
    const monthField = wrapper.find('input[name="monthInputName"]');
    const yearField = wrapper.find('input[name="yearInputName"]');

    dayField.simulate('change', { target: { value: '02' } });
    monthField.simulate('change', { target: { value: '' } });
    yearField.simulate('change', { target: { value: '' } });

    expect(wrapper.find('DateFieldWithPicker').state('selectedDate')).toBe(null);
    expect(mockOnDateChange).toHaveBeenCalledWith({
      target: {
        name: undefined,
        value: 'Invalid date',
      },
    });  });

  it('will set selectedDate to the value entered and set the correct onDateChange prop if the date in the field is complete', ()=> {
    const dayField = wrapper.find('input[name="dayInputName"]');
    const monthField = wrapper.find('input[name="monthInputName"]');
    const yearField = wrapper.find('input[name="yearInputName"]');

    dayField.simulate('change', { target: { value: '10' } });
    monthField.simulate('change', { target: { value: '05' } });
    yearField.simulate('change', { target: { value: '2020' } });

    expect(mockOnDateChange).toHaveBeenCalledWith({
      target: {
        name: undefined,
        value: '2020-05-10',
      },
    });
    expect(wrapper.find('DateFieldWithPicker').state('selectedDate')).not.toBeNull();
  });
});

describe('ComponentDidUpdate', ()=> {

  it('should componentDidUpdate with new props', () => {
    let wrapper;

    const props = {
      errors: undefined,
      date: '2020-05-10',
      dateFormat:'YYYY-MM-DD'
    };

    const prevProps = {
      errors: undefined,
      date: '2020-05-12',
      dateFormat:'YYYY-MM-DD'
    };

    wrapper = mount(<DateFieldWithPicker {...props} />);
    const instance = wrapper.find('DateFieldWithPicker').instance();
    instance.componentDidUpdate(prevProps);

    expect(wrapper.find('DateFieldWithPicker').state('dateObject')).toBeDefined();
    expect(wrapper.find('DateFieldWithPicker').state('dateObject')).toEqual({
      day: '10',
      month: '05',
      year: '2020',
    });
  });


  it('should update dateObject to default value when we get empty date prop', () => {
    let wrapper;
    const props = {
      errors: undefined,
      date: null,
    };

    const prevProps = {
      errors: undefined,
      date: '2020-05-12',
    };

    wrapper = mount(<DateFieldWithPicker {...props} />);

    const instance = wrapper.find('DateFieldWithPicker').instance();
    instance.componentDidUpdate(prevProps);

    expect(wrapper.find('DateFieldWithPicker').state('dateObject')).toBeDefined();
    expect(wrapper.find('DateFieldWithPicker').state('dateObject')).toEqual({
      day: '',
      month: '',
      year: '',
    });
  });
});

describe('ComponentDidMount', ()=> {

  it('should componentDidMount with props and an invalid date', () => {
    let wrapper;

    const props = {
      errors: undefined,
      date: 'Invalid date',
      dateFormat:'YYYY-MM-DD'
    };

    wrapper = mount(<DateFieldWithPicker {...props} />);
    const instance = wrapper.find('DateFieldWithPicker').instance();
    instance.componentDidMount(props);

    expect(wrapper.find('DateFieldWithPicker').state('dateObject')).toBeDefined();
    expect(wrapper.find('DateFieldWithPicker').state('dateObject')).toEqual({
      day: '',
      month: '',
      year: '',
    });
  });

  it('should componentDidMount with props and a correct date', () => {
    let wrapper;

    const props = {
      errors: undefined,
      date: '2020-05-05',
      dateFormat:'YYYY-MM-DD'
    };

    wrapper = mount(<DateFieldWithPicker {...props} />);
    const instance = wrapper.find('DateFieldWithPicker').instance();
    instance.componentDidMount(props);

    expect(wrapper.find('DateFieldWithPicker').state('dateObject')).toBeDefined();
    expect(wrapper.find('DateFieldWithPicker').state('dateObject')).toEqual({
      day: '05',
      month: '05',
      year: '2020',
    });
  });
});

describe('#parseDate', () => {
  it('Should return null if param is not a valid date', () => {
    expect(parseDate('no')).toBeNull();
  });

  it('Should return a parsed date', () => {
    const result = parseDate(new Date('2014-02-27T10:00:00'));
    expect(result).not.toBeNull();
  });
});

describe('#getFormattedDate', () => {
  it('Should return the correctly formatted date', () => {
    const result = getFormattedDate({ day: '01', month: '03', year: '2020' }, 'YYYY-MM-DD');
    expect(result).toBe('2020-03-01');
  });

  it('Should return the correctly formatted date, second option', () => {
    const result = getFormattedDate({ day: '01', month: '03', year: '2020' }, 'DD/MM/YYYY');
    expect(result).toBe('01/03/2020');
  });
});

describe('English/Welsh Translation test cases', () => {
  const props = {
    errors: undefined,
    onDateChange: jest.fn(),
    labelText: '',
    labelTextClass: 'label-landings-form',
    dateFormat:'YYYY-MM-DD',
    id: 'component-id',
    t: jest.fn()
  };
  const wrapper = mount(
      <DateFieldWithPicker {...props} />
  );
});
