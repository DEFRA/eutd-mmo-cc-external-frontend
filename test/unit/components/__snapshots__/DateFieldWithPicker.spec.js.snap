// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DateFieldWithPicker it should create snap shot for whole page 1`] = `
<div>
  <div
    class="date-field-with-picker"
    id="component-id"
  >
    <div
      class="column-field-picker"
    >
      <div
        class="date-field e1vcssft0 css-vav5a4-StyledContainer-StyledHoc e4u1u2g0"
      >
        <span
          class="e1vcssft0 css-3e01s4-StyledLabelText-StyledHoc e4u1u2g0"
        >
          <span
            class="label-landings-form e1vcssft0 css-3e01s4-StyledLabelText-StyledHoc e4u1u2g0"
            id="date-field-label-text"
          />
        </span>
        <span
          class="e1vcssft0 css-1whl0n4-StyledHint-StyledHoc e4u1u2g0"
        >
          For example, 31 03 1980
        </span>
        <div
          class="css-bw34j1-StyledList e107rt5v2"
        >
          <label
            class="e107rt5v1 css-iwpl6b-StyledLabel-StyledHoc-StyledLabel e4u1u2g0"
          >
            <span
              class="e1vcssft0 css-3e01s4-StyledLabelText-StyledHoc e4u1u2g0"
            >
              Day
            </span>
            <input
              class="css-lktfuz-StyledInput e107rt5v0"
              name="dayInputName"
              type="number"
              value=""
            />
          </label>
          <label
            class="e107rt5v1 css-iwpl6b-StyledLabel-StyledHoc-StyledLabel e4u1u2g0"
          >
            <span
              class="e1vcssft0 css-3e01s4-StyledLabelText-StyledHoc e4u1u2g0"
            >
              Month
            </span>
            <input
              class="css-lktfuz-StyledInput e107rt5v0"
              name="monthInputName"
              type="number"
              value=""
            />
          </label>
          <label
            class="e107rt5v1 css-17bpbza-StyledLabel-StyledHoc-StyledLabel e4u1u2g0"
          >
            <span
              class="e1vcssft0 css-3e01s4-StyledLabelText-StyledHoc e4u1u2g0"
            >
              Year
            </span>
            <input
              class="css-lktfuz-StyledInput e107rt5v0"
              name="yearInputName"
              type="number"
              value=""
            />
          </label>
        </div>
      </div>
      <div
        class="date-picker "
      >
        <button>
          <img
            alt="calendar"
            src="/static/assets/images/calendar.png"
          />
        </button>
      </div>
    </div>
  </div>
</div>
`;
